File: shims/cpp-proton/src/receiver.cpp
/*
 * QIT C++ Proton Shim - Receiver Implementation
 */

#include "qit_shim.hpp"
#include <json/json.h>
#include <proton/delivery.hpp>
#include <proton/transport.hpp>
#include <proton/work_queue.hpp>
#include <proton/annotation_key.hpp>
#include <proton/symbol.hpp>
#include <proton/message_id.hpp>
#include <proton/codec/encoder.hpp>
#include <proton/codec/decoder.hpp>
#include <iostream>
#include <vector>
#include <cstdlib>
#include <cstring>
#include <cstdio>
#include <cstdint>

namespace qit {

Receiver::Receiver(const std::string& broker_url,
                   const std::string& queue_name,
                   size_t count,
                   int timeout_sec)
    : broker_url_(broker_url),
      queue_name_(queue_name),
      expected_count_(count),
      received_count_(0),
      timeout_sec_(timeout_sec),
      output_sent_(false) {}

void Receiver::on_container_start(proton::container& c) {
    c.open_receiver(broker_url_ + "/" + queue_name_);

    // Set timeout using std::function wrapped work
    if (timeout_sec_ > 0) {
        proton::work timeout_work = proton::make_work([this]() {
            this->on_timeout();
        });
        c.schedule(proton::duration(timeout_sec_ * 1000), timeout_work);
    }
}

void Receiver::on_message(proton::delivery& d, proton::message& m) {
    try {
        // Check for JMS message type annotation
        // NOTE: Qpid JMS Client uses symbol as key
        int8_t jms_msg_type = -1;
        proton::annotation_key jms_key(proton::symbol("x-opt-jms-msg-type"));
        if (m.message_annotations().exists(jms_key)) {
            proton::value jms_value = m.message_annotations().get(jms_key);
            jms_msg_type = proton::get<int8_t>(jms_value);
        }

        Json::Value decoded;
        if (jms_msg_type >= 0) {
            // Decode as JMS message
            decoded = decode_jms_message(m.body(), jms_msg_type);
        } else {
            // Decode as regular AMQP message
            decoded = TypeCodec::decode(m.body());
        }

        Json::Value msg_data;
        msg_data["index"] = static_cast<int>(received_count_);
        msg_data["type"] = decoded["type"];
        msg_data["value"] = decoded["value"];

        // Extract JMS headers
        Json::Value headers(Json::objectValue);
        try {
            proton::message_id cid = m.correlation_id();
            proton::type_id cid_type = cid.type();
            if (cid_type == proton::BINARY) {
                proton::binary bin = proton::get<proton::binary>(cid);
                Json::Value cid_obj;
                cid_obj["type"] = "bytes";
                cid_obj["value"] = binary_to_hex(bin);
                headers["JMSCorrelationID"] = cid_obj;
            } else if (cid_type == proton::STRING) {
                headers["JMSCorrelationID"] = proton::get<std::string>(cid);
            }
        } catch (...) {}

        std::string reply_to = m.reply_to();
        if (!reply_to.empty()) {
            Json::Value rt_obj;
            std::string reply_type = "queue";
            proton::annotation_key rt_key(proton::symbol("x-opt-jms-reply-to"));
            if (m.message_annotations().exists(rt_key)) {
                proton::value rt_val = m.message_annotations().get(rt_key);
                if (proton::get<int8_t>(rt_val) == 1) reply_type = "topic";
            } else if (reply_to.substr(0, 8) == "topic://") {
                reply_type = "topic";
                reply_to = reply_to.substr(8);
            } else if (reply_to.substr(0, 8) == "queue://") {
                reply_to = reply_to.substr(8);
            }
            rt_obj["type"] = reply_type;
            rt_obj["value"] = reply_to;
            headers["JMSReplyTo"] = rt_obj;
        }

        std::string subject = m.subject();
        if (!subject.empty()) {
            headers["JMSType"] = subject;
        }

        if (headers.size() > 0) {
            msg_data["headers"] = headers;
        }

        // Extract application properties
        Json::Value props(Json::objectValue);
        try {
            if (!m.properties().empty()) {
                const proton::value& pval = m.properties().value();
                proton::codec::decoder dec(pval);
                proton::codec::start s;
                dec >> s;
                for (size_t pi = 0; pi < s.size / 2; ++pi) {
                    std::string key;
                    proton::scalar val;
                    dec >> key >> val;

                    Json::Value prop_obj;
                    proton::type_id tid = val.type();
                    char hex_buf[32];

                    if (tid == proton::BOOLEAN) {
                        prop_obj["type"] = "boolean";
                        prop_obj["value"] = proton::get<bool>(val);
                    } else if (tid == proton::BYTE) {
                        prop_obj["type"] = "byte";
                        int8_t v = proton::get<int8_t>(val);
                        snprintf(hex_buf, sizeof(hex_buf), "0x%02x", static_cast<unsigned int>(v & 0xFF));
                        prop_obj["value"] = std::string(hex_buf);
                    } else if (tid == proton::SHORT) {
                        prop_obj["type"] = "short";
                        int16_t v = proton::get<int16_t>(val);
                        snprintf(hex_buf, sizeof(hex_buf), "0x%04x", static_cast<unsigned int>(v & 0xFFFF));
                        prop_obj["value"] = std::string(hex_buf);
                    } else if (tid == proton::INT) {
                        prop_obj["type"] = "int";
                        int32_t v = proton::get<int32_t>(val);
                        snprintf(hex_buf, sizeof(hex_buf), "0x%08x", static_cast<unsigned int>(v));
                        prop_obj["value"] = std::string(hex_buf);
                    } else if (tid == proton::LONG) {
                        prop_obj["type"] = "long";
                        int64_t v = proton::get<int64_t>(val);
                        snprintf(hex_buf, sizeof(hex_buf), "0x%016llx", static_cast<unsigned long long>(v));
                        prop_obj["value"] = std::string(hex_buf);
                    } else if (tid == proton::FLOAT) {
                        prop_obj["type"] = "float";
                        float fv = proton::get<float>(val);
                        uint32_t bits;
                        std::memcpy(&bits, &fv, sizeof(bits));
                        snprintf(hex_buf, sizeof(hex_buf), "0x%08x", bits);
                        prop_obj["value"] = std::string(hex_buf);
                    } else if (tid == proton::DOUBLE) {
                        prop_obj["type"] = "double";
                        double dv = proton::get<double>(val);
                        uint64_t bits;
                        std::memcpy(&bits, &dv, sizeof(bits));
                        snprintf(hex_buf, sizeof(hex_buf), "0x%016llx", static_cast<unsigned long long>(bits));
                        prop_obj["value"] = std::string(hex_buf);
                    } else if (tid == proton::STRING) {
                        prop_obj["type"] = "string";
                        prop_obj["value"] = proton::get<std::string>(val);
                    } else {
                        continue;
                    }

                    props[key] = prop_obj;
                }
                dec >> proton::codec::finish();
            }
        } catch (...) {}

        if (props.size() > 0) {
            msg_data["properties"] = props;
        }

        received_messages_.append(msg_data);
        received_count_++;

        if (received_count_ >= expected_count_) {
            d.receiver().close();
            d.connection().close();

            // Output result
            output_result();
        }
    } catch (const std::exception& e) {
        std::cerr << "Error processing message: " << e.what() << std::endl;
        d.receiver().close();
        d.connection().close();
        throw;
    }
}

Json::Value Receiver::decode_jms_message(const proton::value& body, int8_t jms_msg_type) {
    // JMS message type constants
    const int8_t JMS_MESSAGE = 0;
    const int8_t JMS_TEXT_MESSAGE = 5;
    const int8_t JMS_BYTES_MESSAGE = 3;
    const int8_t JMS_MAP_MESSAGE = 2;
    const int8_t JMS_STREAM_MESSAGE = 4;

    Json::Value result;

    if (jms_msg_type == JMS_TEXT_MESSAGE) {
        // TextMessage: body is string in AmqpValue section
        result["type"] = "text";  // Use 'text' to match JMS shim output
        try {
            result["value"] = proton::get<std::string>(body);
        } catch (...) {
            result["value"] = Json::nullValue;
        }
    } else if (jms_msg_type == JMS_BYTES_MESSAGE) {
        // BytesMessage: body is binary in Data section
        result["type"] = "bytes";
        try {
            proton::binary bin = proton::get<proton::binary>(body);
            std::string hex;
            for (uint8_t byte : bin) {
                char buf[3];
                snprintf(buf, sizeof(buf), "%02x", byte);
                hex += buf;
            }
            result["value"] = hex;
        } catch (...) {
            result["value"] = Json::nullValue;
        }
    } else if (jms_msg_type == JMS_MESSAGE) {
        // Empty message
        result["type"] = "null";
        result["value"] = Json::nullValue;
    } else if (jms_msg_type == JMS_MAP_MESSAGE) {
        try {
            proton::codec::decoder dec(body);
            proton::codec::start s;
            dec >> s;
            if (s.size >= 2) {
                std::string key;
                proton::value val;
                dec >> key >> val;
                dec >> proton::codec::finish();
                Json::Value decoded = TypeCodec::decode(val);
                result["type"] = decoded["type"];
                result["value"] = decoded["value"];
            } else {
                result["type"] = "none";
                result["value"] = Json::nullValue;
            }
        } catch (...) {
            result["type"] = "none";
            result["value"] = Json::nullValue;
        }
    } else if (jms_msg_type == JMS_STREAM_MESSAGE) {
        try {
            proton::codec::decoder dec(body);
            proton::codec::start s;
            dec >> s;
            if (s.size >= 1) {
                proton::value val;
                dec >> val;
                dec >> proton::codec::finish();
                Json::Value decoded = TypeCodec::decode(val);
                result["type"] = decoded["type"];
                result["value"] = decoded["value"];
            } else {
                result["type"] = "none";
                result["value"] = Json::nullValue;
            }
        } catch (...) {
            result["type"] = "none";
            result["value"] = Json::nullValue;
        }
    } else {
        // Unknown JMS type, fall back to regular AMQP decoding
        return TypeCodec::decode(body);
    }

    return result;
}

void Receiver::on_timeout() {
    // Output what we received so far
    output_result();

    if (received_count_ < expected_count_) {
        std::exit(1);  // Exit with error if we didn't get all messages
    }
}

void Receiver::output_result() {
    if (output_sent_) return;  // Already output, don't duplicate
    output_sent_ = true;

    Json::Value result;
    result["messages"] = received_messages_;
    result["stats"]["received"] = static_cast<Json::Value::UInt>(received_count_);

    Json::StreamWriterBuilder builder;
    builder["indentation"] = "  ";
    std::cout << Json::writeString(builder, result) << std::endl;
}

void Receiver::on_transport_error(proton::transport& t) {
    std::cerr << "Transport error: " << t.error() << std::endl;
}

void Receiver::on_error(const proton::error_condition& ec) {
    std::cerr << "Error: " << ec << std::endl;
}

// --- Large Content Receiver ---

LargeContentReceiver::LargeContentReceiver(const std::string& broker_url,
                                           const std::string& queue_name,
                                           const std::string& content_type,
                                           uint32_t seed,
                                           size_t size,
                                           int timeout_sec)
    : broker_url_(broker_url),
      queue_name_(queue_name),
      content_type_(content_type),
      seed_(seed),
      size_(size),
      timeout_sec_(timeout_sec),
      received_(false) {}

void LargeContentReceiver::on_container_start(proton::container& c) {
    c.open_receiver(broker_url_ + "/" + queue_name_);

    if (timeout_sec_ > 0) {
        proton::work timeout_work = proton::make_work([this]() {
            this->on_timeout();
        });
        c.schedule(proton::duration(timeout_sec_ * 1000), timeout_work);
    }
}

void LargeContentReceiver::on_message(proton::delivery& d, proton::message& m) {
    if (received_) return;
    received_ = true;

    Json::Value result;

    try {
        if (content_type_ == "binary") {
            proton::binary received_bin = proton::get<proton::binary>(m.body());
            auto expected = lcg_generate_bytes(seed_, size_);

            size_t received_size = received_bin.size();
            result["size"] = static_cast<Json::Value::UInt64>(received_size);
            result["expected_size"] = static_cast<Json::Value::UInt64>(size_);

            if (received_size != size_) {
                result["match"] = false;
            } else {
                bool match = true;
                for (size_t i = 0; i < size_; i++) {
                    if (static_cast<uint8_t>(received_bin[i]) != expected[i]) {
                        result["match"] = false;
                        result["first_mismatch_offset"] = static_cast<Json::Value::UInt64>(i);
                        match = false;
                        break;
                    }
                }
                if (match) {
                    result["match"] = true;
                }
            }
        } else {
            std::string received_str = proton::get<std::string>(m.body());
            std::string expected = lcg_generate_string(seed_, size_);

            size_t received_size = received_str.size();
            result["size"] = static_cast<Json::Value::UInt64>(received_size);
            result["expected_size"] = static_cast<Json::Value::UInt64>(size_);

            if (received_size != size_) {
                result["match"] = false;
            } else if (received_str == expected) {
                result["match"] = true;
            } else {
                result["match"] = false;
                for (size_t i = 0; i < size_; i++) {
                    if (received_str[i] != expected[i]) {
                        result["first_mismatch_offset"] = static_cast<Json::Value::UInt64>(i);
                        break;
                    }
                }
            }
        }
    } catch (const std::exception& e) {
        result["match"] = false;
        result["error"] = std::string("Failed to extract body: ") + e.what();
    }

    Json::StreamWriterBuilder builder;
    builder["indentation"] = "  ";
    std::cout << Json::writeString(builder, result) << std::endl;

    d.receiver().close();
    d.connection().close();
}

void LargeContentReceiver::on_timeout() {
    if (!received_) {
        Json::Value result;
        result["match"] = false;
        result["error"] = "no message received";

        Json::StreamWriterBuilder builder;
        builder["indentation"] = "  ";
        std::cout << Json::writeString(builder, result) << std::endl;

        std::exit(1);
    }
}

void LargeContentReceiver::on_transport_error(proton::transport& t) {
    std::cerr << "Transport error: " << t.error() << std::endl;
}

} // namespace qit

File: shims/dotnet-proton/src/Receiver.cs
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/*
 * QIT .NET Apache Qpid Proton Shim - Receiver
 */

using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using Apache.Qpid.Proton.Buffer;
using Apache.Qpid.Proton.Client;
using Apache.Qpid.Proton.Types;
using Apache.Qpid.Proton.Types.Messaging;
using Newtonsoft.Json;

namespace Qit.Shim
{
    public static class Receiver
    {
        public static void Receive(string broker, string queue, int count, int timeout)
        {
            try
            {
                var messages = new List<MessageResult>();

                // Parse broker URL
                var brokerUri = ParseBrokerUrl(broker);

                // Create client and connect
                IClient client = IClient.Create();

                ConnectionOptions options = new ConnectionOptions
                {
                    User = "artemis",
                    Password = "artemis"
                };

                using IConnection connection = client.Connect(brokerUri.Host, brokerUri.Port, options);
                using IReceiver receiver = connection.OpenReceiver(queue);

                // Receive messages with timeout
                var timeoutMs = timeout * 1000;
                for (int i = 0; i < count; i++)
                {
                    IDelivery delivery = receiver.Receive(TimeSpan.FromMilliseconds(timeoutMs));

                    if (delivery == null)
                    {
                        break;  // Timeout
                    }

                    IMessage<object> message = delivery.Message();

                    // Check for JMS message type annotation
                    sbyte jmsType = -1;
                    if (message.HasAnnotation("x-opt-jms-msg-type"))
                    {
                        jmsType = Convert.ToSByte(message.GetAnnotation("x-opt-jms-msg-type"));
                    }

                    DecodedMessage decoded;
                    if (jmsType >= 0)
                    {
                        // Decode as JMS message
                        decoded = DecodeJmsMessage(message.Body, jmsType);
                    }
                    else
                    {
                        bool isAmqpValue = false;
                        var advanced = message as IAdvancedMessage<object>;
                        if (advanced != null)
                        {
                            var sections = advanced.GetBodySections();
                            if (sections != null)
                            {
                                foreach (var section in sections)
                                {
                                    if (section is AmqpValue)
                                        isAmqpValue = true;
                                    break;
                                }
                            }
                        }
                        decoded = TypeCodec.Decode(message.Body, isAmqpValue);
                    }

                    var msgResult = new MessageResult
                    {
                        Index = i,
                        Type = decoded.Type,
                        Value = decoded.Value
                    };

                    // Extract JMS headers
                    var hdrs = new Dictionary<string, object>();
                    if (message.CorrelationId != null)
                    {
                        if (message.CorrelationId is byte[] corrBytes)
                        {
                            hdrs["JMSCorrelationID"] = new Dictionary<string, string>
                            {
                                { "type", "bytes" },
                                { "value", BitConverter.ToString(corrBytes).Replace("-", "").ToLower() }
                            };
                        }
                        else if (message.CorrelationId is IProtonBuffer buf)
                        {
                            var bytes = new byte[buf.ReadableBytes];
                            buf.CopyInto(buf.ReadOffset, bytes, 0, bytes.Length);
                            hdrs["JMSCorrelationID"] = new Dictionary<string, string>
                            {
                                { "type", "bytes" },
                                { "value", BitConverter.ToString(bytes).Replace("-", "").ToLower() }
                            };
                        }
                        else
                        {
                            hdrs["JMSCorrelationID"] = message.CorrelationId.ToString();
                        }
                    }
                    if (message.ReplyTo != null)
                    {
                        string replyType = "queue";
                        string replyAddr = message.ReplyTo;
                        if (message.HasAnnotation("x-opt-jms-reply-to"))
                        {
                            var rt = Convert.ToSByte(message.GetAnnotation("x-opt-jms-reply-to"));
                            if (rt == 1) replyType = "topic";
                        }
                        else if (replyAddr.StartsWith("topic://"))
                        {
                            replyType = "topic";
                            replyAddr = replyAddr.Substring(8);
                        }
                        else if (replyAddr.StartsWith("queue://"))
                        {
                            replyAddr = replyAddr.Substring(8);
                        }
                        hdrs["JMSReplyTo"] = new Dictionary<string, string>
                        {
                            { "type", replyType },
                            { "value", replyAddr }
                        };
                    }
                    if (message.Subject != null)
                    {
                        hdrs["JMSType"] = message.Subject;
                    }
                    if (hdrs.Count > 0)
                    {
                        msgResult.Headers = hdrs;
                    }

                    // Extract JMS application properties
                    var props = new Dictionary<string, object>();
                    message.ForEachProperty((name, value) =>
                    {
                        if (value is bool b)
                        {
                            var boolProp = new Dictionary<string, object>
                            {
                                { "type", "boolean" },
                                { "value", b }
                            };
                            props[name] = boolProp;
                            return;
                        }
                        Dictionary<string, string> prop;
                        if (value is sbyte sb)
                        {
                            prop = new Dictionary<string, string>
                            {
                                { "type", "byte" },
                                { "value", $"0x{(sb & 0xFF):x2}" }
                            };
                        }
                        else if (value is short s)
                        {
                            prop = new Dictionary<string, string>
                            {
                                { "type", "short" },
                                { "value", $"0x{(s & 0xFFFF):x4}" }
                            };
                        }
                        else if (value is int i)
                        {
                            prop = new Dictionary<string, string>
                            {
                                { "type", "int" },
                                { "value", $"0x{(uint)i:x8}" }
                            };
                        }
                        else if (value is long l)
                        {
                            prop = new Dictionary<string, string>
                            {
                                { "type", "long" },
                                { "value", $"0x{(ulong)l:x16}" }
                            };
                        }
                        else if (value is float f)
                        {
                            var bits = BitConverter.SingleToInt32Bits(f);
                            prop = new Dictionary<string, string>
                            {
                                { "type", "float" },
                                { "value", $"0x{(uint)bits:x8}" }
                            };
                        }
                        else if (value is double d)
                        {
                            var bits = BitConverter.DoubleToInt64Bits(d);
                            prop = new Dictionary<string, string>
                            {
                                { "type", "double" },
                                { "value", $"0x{(ulong)bits:x16}" }
                            };
                        }
                        else if (value is string str)
                        {
                            prop = new Dictionary<string, string>
                            {
                                { "type", "string" },
                                { "value", str }
                            };
                        }
                        else
                        {
                            prop = new Dictionary<string, string>
                            {
                                { "type", "string" },
                                { "value", value?.ToString() ?? "" }
                            };
                        }
                        props[name] = prop;
                    });
                    if (props.Count > 0)
                    {
                        msgResult.Properties = props;
                    }

                    // Extract AMQP Header section fields
                    msgResult.MessageHeader = new Dictionary<string, object>
                    {
                        { "durable", message.Durable },
                        { "priority", (int)message.Priority },
                        { "ttl", (long)message.TimeToLive },
                        { "first_acquirer", message.FirstAcquirer },
                        { "delivery_count", (int)message.DeliveryCount },
                    };

                    messages.Add(msgResult);
                }

                // Output result
                var result = new
                {
                    messages,
                    stats = new { received = messages.Count }
                };

                Console.WriteLine(JsonConvert.SerializeObject(result, Formatting.Indented));

                if (messages.Count < count)
                {
                    Environment.Exit(1);
                }
            }
            catch (Exception ex)
            {
                Console.Error.WriteLine($"Receive error: {ex.Message}");
                Environment.Exit(1);
            }
        }

        public static void ReceiveLargeContent(string broker, string queue, string contentType, int size, int seed, int timeout, int elements = 0, int elementSize = 0)
        {
            try
            {
                var brokerUri = ParseBrokerUrl(broker);

                IClient client = IClient.Create();
                ConnectionOptions options = new ConnectionOptions
                {
                    User = "artemis",
                    Password = "artemis"
                };

                using IConnection connection = client.Connect(brokerUri.Host, brokerUri.Port, options);
                using IReceiver receiver = connection.OpenReceiver(queue);

                IDelivery delivery = receiver.Receive(TimeSpan.FromSeconds(timeout));
                if (delivery == null)
                {
                    Console.WriteLine(JsonConvert.SerializeObject(new { match = false, error = "no message received" }));
                    Environment.Exit(1);
                }

                IMessage<object> message = delivery.Message();

                object receivedBody = message.Body;
                bool matched;
                int receivedSize;
                int? firstMismatchOffset = null;

                if (contentType == "binary")
                {
                    byte[] expected = LcgHelper.LcgGenerateBytes((uint)seed, size);
                    byte[] received;
                    if (receivedBody is byte[] byteArray)
                        received = byteArray;
                    else if (receivedBody is IProtonBuffer buf)
                    {
                        received = new byte[buf.ReadableBytes];
                        buf.CopyInto(buf.ReadOffset, received, 0, received.Length);
                    }
                    else
                    {
                        Console.WriteLine(JsonConvert.SerializeObject(new { match = false, error = "expected binary body but got " + (receivedBody?.GetType().Name ?? "null") }));
                        Environment.Exit(1);
                        return;
                    }

                    receivedSize = received.Length;
                    if (received.Length != expected.Length)
                    {
                        matched = false;
                    }
                    else
                    {
                        matched = true;
                        for (int i = 0; i < expected.Length; i++)
                        {
                            if (received[i] != expected[i])
                            {
                                matched = false;
                                firstMismatchOffset = i;
                                break;
                            }
                        }
                    }
                }
                else if (contentType == "string")
                {
                    string expected = LcgHelper.LcgGenerateString((uint)seed, size);
                    string received = receivedBody as string ?? receivedBody?.ToString() ?? "";

                    receivedSize = received.Length;
                    if (received.Length != expected.Length)
                    {
                        matched = false;
                    }
                    else
                    {
                        matched = true;
                        for (int i = 0; i < expected.Length; i++)
                        {
                            if (received[i] != expected[i])
                            {
                                matched = false;
                                firstMismatchOffset = i;
                                break;
                            }
                        }
                    }
                }
                else if (contentType == "list" || contentType == "array" || contentType == "map" || contentType == "described")
                {
                    var expected = LcgHelper.GenerateCollectionElements((uint)seed, elements, elementSize);
                    var received = new List<string>();

                    if (contentType == "list")
                    {
                        if (receivedBody is IList list)
                        {
                            foreach (var item in list)
                                received.Add(item?.ToString() ?? "");
                        }
                        else
                        {
                            Console.WriteLine(JsonConvert.SerializeObject(new { match = false, error = "expected list, got " + (receivedBody?.GetType().Name ?? "null") }));
                            Environment.Exit(1);
                            return;
                        }
                    }
                    else if (contentType == "array")
                    {
                        if (receivedBody is string[] strArr)
                        {
                            received.AddRange(strArr);
                        }
                        else if (receivedBody is object[] objArr)
                        {
                            foreach (var o in objArr) received.Add(o?.ToString() ?? "");
                        }
                        else if (receivedBody is IList arrList)
                        {
                            foreach (var item in arrList) received.Add(item?.ToString() ?? "");
                        }
                        else
                        {
                            Console.WriteLine(JsonConvert.SerializeObject(new { match = false, error = "expected array, got " + (receivedBody?.GetType().Name ?? "null") }));
                            Environment.Exit(1);
                            return;
                        }
                    }
                    else if (contentType == "map")
                    {
                        if (receivedBody is IDictionary dict)
                        {
                            var keys = LcgHelper.GenerateMapKeys(elements);
                            foreach (var key in keys)
                            {
                                received.Add(dict.Contains(key) ? dict[key]?.ToString() ?? "" : "");
                            }
                        }
                        else
                        {
                            Console.WriteLine(JsonConvert.SerializeObject(new { match = false, error = "expected map, got " + (receivedBody?.GetType().Name ?? "null") }));
                            Environment.Exit(1);
                            return;
                        }
                    }
                    else if (contentType == "described")
                    {
                        object inner = receivedBody;
                        // Check if it's a described type (IDescribedType) and unwrap
                        if (receivedBody is IDescribedType desc)
                            inner = desc.Described;

                        if (inner is IList descList)
                        {
                            foreach (var item in descList) received.Add(item?.ToString() ?? "");
                        }
                        else
                        {
                            Console.WriteLine(JsonConvert.SerializeObject(new { match = false, error = "expected described list, got " + (inner?.GetType().Name ?? "null") }));
                            Environment.Exit(1);
                            return;
                        }
                    }

                    // Compare element by element
                    var collResult = new Dictionary<string, object>
                    {
                        { "elements", received.Count },
                        { "element_size", elementSize }
                    };

                    if (received.Count != elements)
                    {
                        collResult["match"] = false;
                    }
                    else
                    {
                        bool collMatched = true;
                        for (int i = 0; i < elements; i++)
                        {
                            if (expected[i] != received[i])
                            {
                                collMatched = false;
                                collResult["first_mismatch_element"] = i;
                                int minLen = Math.Min(expected[i].Length, received[i].Length);
                                int offset = minLen;
                                for (int j = 0; j < minLen; j++)
                                {
                                    if (expected[i][j] != received[i][j])
                                    {
                                        offset = j;
                                        break;
                                    }
                                }
                                collResult["first_mismatch_offset"] = offset;
                                break;
                            }
                        }
                        collResult["match"] = collMatched;
                    }

                    Console.WriteLine(JsonConvert.SerializeObject(collResult));
                    if (!(bool)collResult["match"])
                        Environment.Exit(1);
                    return;
                }
                else
                {
                    Console.WriteLine(JsonConvert.SerializeObject(new { match = false, error = $"unknown content type: {contentType}" }));
                    Environment.Exit(1);
                    return;
                }

                var result = new Dictionary<string, object>
                {
                    { "match", matched },
                    { "size", receivedSize },
                    { "expected_size", size }
                };
                if (firstMismatchOffset.HasValue)
                    result["first_mismatch_offset"] = firstMismatchOffset.Value;

                Console.WriteLine(JsonConvert.SerializeObject(result));

                if (!matched)
                    Environment.Exit(1);
            }
            catch (Exception ex)
            {
                Console.Error.WriteLine($"Receive error: {ex.Message}");
                Environment.Exit(1);
            }
        }

        private static (string Host, int Port) ParseBrokerUrl(string broker)
        {
            var uri = new Uri(broker.StartsWith("amqp://") ? broker : $"amqp://{broker}");
            return (uri.Host, uri.Port > 0 ? uri.Port : 5672);
        }

        private static DecodedMessage DecodeJmsMessage(object body, sbyte jmsType)
        {
            // JMS message type constants
            const sbyte JMS_MESSAGE = 0;
            const sbyte JMS_TEXT_MESSAGE = 5;
            const sbyte JMS_BYTES_MESSAGE = 3;
            const sbyte JMS_MAP_MESSAGE = 2;
            const sbyte JMS_STREAM_MESSAGE = 4;

            if (jmsType == JMS_TEXT_MESSAGE)
            {
                // TextMessage: body is string in AmqpValue section
                return new DecodedMessage
                {
                    Type = "text",  // Use 'text' to match JMS shim output
                    Value = body as string
                };
            }
            else if (jmsType == JMS_BYTES_MESSAGE)
            {
                // BytesMessage: body is binary in Data section
                if (body is byte[] bytes)
                {
                    return new DecodedMessage
                    {
                        Type = "bytes",
                        Value = BitConverter.ToString(bytes).Replace("-", "").ToLower()
                    };
                }
                return new DecodedMessage { Type = "bytes", Value = "" };
            }
            else if (jmsType == JMS_MESSAGE)
            {
                // Empty message
                return new DecodedMessage
                {
                    Type = "null",
                    Value = null
                };
            }

            if (jmsType == JMS_MAP_MESSAGE)
            {
                if (body is IDictionary dict && dict.Count > 0)
                {
                    var enumerator = dict.GetEnumerator();
                    enumerator.MoveNext();
                    return TypeCodec.Decode(enumerator.Value);
                }
                return new DecodedMessage { Type = "none", Value = null };
            }

            if (jmsType == JMS_STREAM_MESSAGE)
            {
                if (body is IList list && list.Count > 0)
                {
                    return TypeCodec.Decode(list[0]);
                }
                return new DecodedMessage { Type = "none", Value = null };
            }

            // Unknown JMS type, fall back to regular AMQP decoding
            return TypeCodec.Decode(body);
        }
    }
}
